### PR TITLE
Added dstat_nvidia_gpu plugin which reports gpu usage.

### PR DESCRIPTION
**Pull request by datumbox**
##### ISSUE TYPE
 - New plugin pull-request

##### DSTAT VERSION
```
Dstat 0.7.2
Written by Dag Wieers <dag@wieers.com>
Homepage at http://dag.wieers.com/home-made/dstat/

Platform posix/linux2
Kernel 4.4.0-62-generic
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609]

Terminal type: xterm (color support)
Terminal size: 24 lines, 96 columns

Processors: 64
Pagesize: 4096
Clock ticks per secs: 100

internal:
	aio, cpu, cpu24, disk, disk24, disk24old, epoch, fs, int, int24, io, ipc, load, lock, 
	mem, net, page, page24, proc, raw, socket, swap, swapold, sys, tcp, time, udp, unix, 
	vm
/usr/share/dstat:
	battery, battery-remain, cpufreq, dbus, disk-tps, disk-util, dstat, dstat-cpu, 
	dstat-ctxt, dstat-mem, fan, freespace, gpfs, gpfs-ops, helloworld, innodb-buffer, 
	innodb-io, innodb-ops, lustre, memcache-hits, mysql-io, mysql-keys, mysql5-cmds, 
	mysql5-io, mysql5-keys, net-packets, nfs3, nfs3-ops, nfsd3, nfsd3-ops, ntp, 
	nvidia-gpu, postfix, power, proc-count, qmail, rpc, rpcd, sendmail, snooze, squid, 
	test, thermal, top-bio, top-bio-adv, top-childwait, top-cpu, top-cpu-adv, top-cputime, 
	top-cputime-avg, top-int, top-io, top-io-adv, top-latency, top-latency-avg, top-mem, 
	top-oom, utmp, vm-memctl, vmk-hba, vmk-int, vmk-nic, vz-cpu, vz-io, vz-ubc, wifi
```

##### SUMMARY
Created a new plugin that reports the GPU usage for nvidia cards. This would be useful for those that do GPU programming. The plugin requires the [nvidia-ml-py](https://pypi.python.org/pypi/nvidia-ml-py) package to collect data from the graphics cards.

### CODE REVIEW

File: plugins/dstat_nvidia_gpu.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self):
+        self.name = 'gpu nvidia'
+        self.nick = ('gpu',)
+        self.vars = ('gpu',)
+        self.type = 'p'
+        self.width = 5
+        self.scale = 1
+
+    def check(self):
+        import pynvml
```
Comment (dagwieers) [inline plugins/dstat_nvidia_gpu.py]: If the import fails, we have to catch it and provide a proper error to the user to assist them.
```python
        try:
            import pynvml
        except:
            raise Exception, 'Needs python-XXXX module'
```

File: plugins/dstat_nvidia_gpu.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        pynvml.nvmlInit()
+
+    def extract(self):
+        self.val['gpu'] = self._getTotalUsage(10)
+
+    def _getUsagePerGPU(self, samples):
```
Comment (dagwieers) [inline plugins/dstat_nvidia_gpu.py]: I don't know if it is useful (and I let you decide), but if you have per-cpu statistics, maybe some people would like to see them for each of their GPUs.

**Comment by dagwieers** [comment]
Looking good. I don't have the hardware so cannot test.

File: plugins/dstat_nvidia_gpu.py (hunk 1 above)
Comment (dagwieers) [inline plugins/dstat_nvidia_gpu.py]: Just noticed that we already catch the import error in Dstat itself:
```
Module dstat_nvidia_gpu failed to load. (No module named pynvml)
```

So if you consider that good enough, I don't think we have to do something special.
(Maybe I should improve that error to state: `The "pynvml library is missing from this system.`)

Comment (datumbox) [inline plugins/dstat_nvidia_gpu.py]: Hehe... My thoughts exactly. Your code should be able to handle ion the Dstat class. Originally I planned to try-catch it but then I saw that Python 2 and 3 have different way to check if the module exists. Is your code compatible with both python versions?

File: plugins/dstat_nvidia_gpu.py (hunk 2 above)
Comment (datumbox) [inline plugins/dstat_nvidia_gpu.py]: Yeah, happy to do it. Can you point me to the snippet that does the same for the CPU? The base dstat class does not provide comments for how the various internal fields work (vars, type, scale etc) so I'm not sure in which format I should return the results.

**Comment by datumbox** [comment]
@dagwieers: I added the improved error messages as you requested. Also if the -f flag is provided, the plugin will print not only the total but also the usage for each gpu. 

Ideally though the user should be able to configure the plugin the same way as with CPU. Unfortunately, as far as I understand, external plugins in dstat can't receive input parameters. I believe the limitation is imposed by line 117 on the dstat.py file, where only two longopts are allowed to receive parameters (the output and the pidfile). 

If on the future it becomes possible for plugins to receive input parameters I can have a second look. I think I'm happy with how the plugin works for now.

**Comment by datumbox** [comment]
@dagwieers Here is the output of the plugin since you can't test it yourself. :)

```bash
dstat --nvidia-gpu
gpu-u
total
   15
    5
   18
   16
```
```bash
dstat --nvidia-gpu -f
-------------------------------------------gpu-usage-nvidia------------------------------------------
total  gpu0  gpu1  gpu2  gpu3  gpu4  gpu5  gpu6  gpu7  gpu8  gpu9 gpu10 gpu11 gpu12 gpu13 gpu14 gpu15
   14    21    21    15    19    18    18    12    11    14    10    11     9     8     8    11    14
   19    25    23    23    22    22    23    20    24    15    15    17    16    16    17    16    16
   19    24    19    21    25    25    23    23    22    14    15    17    15    18    18    17    14
    7     8     7     8     7     7     8     8     9     3     6     7     7     7     7     5     4
   20    26    21    23    22    23    22    23    24    16    16    17    16    15    16    15    17
   20    22    24    23    26    24    22    25    24    16    17    17    16    18    17    17    16
   15    20    18    20    18    19    17    17    19    13    14    14    14    11    12    10    12
   18    23    20    20    21    23    21    21    21    15    16    15    16    14    16    16    14
    5     7     4     7     4     4     3     9     7     3     5     5     6     6     6     3     3
   13    16    16    14    15    15    15    19    14     9    12    10    12    14    13    11    10
   21    29    24    23    24    24    25    24    25    16    17    18    18    18    16    17    17
   18    20    20    22    22    21    21    20    22    14    15    14    14    15    15    13    15
   15    18    17    16    17    16    15    19    17    11    13    11    13    15    13    13    13
    7     9     7     9     6     7     7     8     9     4     7     6     7     5     5     5     4
```

**Comment by datumbox** [comment]
@dagwieers Do you need any other changes before merging?

**Pull request closed without merging**